rust/lance/src/dataset/scanner.rs

```diff
@@ -825,6 +912,83 @@ impl Scanner {
         Ok(knn_node)
     }
 
+    #[async_recursion]
+    async fn fragments_covered_by_index_query(
+        &self,
+        index_expr: &ScalarIndexExpr,
+    ) -> Result<RoaringBitmap> {
+        match index_expr {
+            ScalarIndexExpr::And(lhs, rhs) => {
+                Ok(self.fragments_covered_by_index_query(lhs).await?
+                    & self.fragments_covered_by_index_query(rhs).await?)
+            }
+            ScalarIndexExpr::Or(lhs, rhs) => Ok(self.fragments_covered_by_index_query(lhs).await?
+                & self.fragments_covered_by_index_query(rhs).await?),
+            ScalarIndexExpr::Not(expr) => self.fragments_covered_by_index_query(expr).await,
+            ScalarIndexExpr::Query(column, _) => {
+                let idx = self
+                    .dataset
+                    .load_scalar_index_for_column(column)
+                    .await?
+                    .expect("Index not found even though it must have been found earlier");
+                Ok(idx
+                    .fragment_bitmap
+                    .expect("scalar indices should always have a fragment bitmap"))
+            }
+        }
+    }
+
+    async fn indexed_scan(
```

Comment:
Is this just for scalar index scan?

Reply:
Yes, I can put scalar in the name.

Reply:
Changed to `scalar_indexed_scan` and added a comment